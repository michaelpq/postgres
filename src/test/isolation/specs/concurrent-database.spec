# Test for interactions with DDL commands manipulating database
# objects.
# XXX: This cannot be included in the final patch as installcheck would
# create a database on existing instances!
# The following set of commands can interact in concurrency:
# - ALTER DATABASE SET
# - ALTER DATABASE OWNER TO
# - ALTER DATABASE SET TABLESPACE
# - ALTER DATABASE WITH

setup
{
  CREATE DATABASE regress_database TEMPLATE template0;
}

teardown
{
  DROP DATABASE regress_database;
}

session "s1"
step "s1_begin"			{ BEGIN; }
step "s1_alterdb_with"	{
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true; }
step "s1_alterdb_set"	{
	ALTER DATABASE regress_database SET commit_delay = '10'; }
step "s1_alterdb_owner"	{
	ALTER DATABASE regress_database OWNER TO CURRENT_USER; }
step "s1_alterdb_tbc"	{
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT; }
step "s1_commit"	{ COMMIT; }

session "s2"
step "s2_begin"			{ BEGIN; }
step "s2_alterdb_with"	{
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true; }
step "s2_alterdb_set"	{
	ALTER DATABASE regress_database SET commit_delay = '10'; }
step "s2_alterdb_owner"	{
	ALTER DATABASE regress_database OWNER TO CURRENT_USER; }
step "s2_alterdb_tbc"	{
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT; }
step "s2_commit"	{ COMMIT; }

permutation "s1_begin" "s1_alterdb_with" "s2_begin" "s2_alterdb_with"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_with" "s2_begin" "s2_alterdb_set"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_with" "s2_begin" "s2_alterdb_owner"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_with" "s2_begin" "s2_alterdb_tbc"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_set" "s2_begin" "s2_alterdb_set"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_set" "s2_begin" "s2_alterdb_owner"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_set" "s2_begin" "s2_alterdb_tbc"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_owner" "s2_begin" "s2_alterdb_owner"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_owner" "s2_begin" "s2_alterdb_tbc"
	"s1_commit" "s2_commit"
permutation "s1_begin" "s1_alterdb_tbc" "s2_begin" "s2_alterdb_tbc"
	"s1_commit" "s2_commit"
