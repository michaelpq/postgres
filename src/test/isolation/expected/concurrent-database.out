Parsed test spec with 2 sessions

starting permutation: s1_begin s1_alterdb_with s2_begin s2_alterdb_with s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_with: 
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true;
step s2_begin: BEGIN;
step s2_alterdb_with: 
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true; <waiting ...>
step s1_commit: COMMIT;
step s2_alterdb_with: <... completed>
error in steps s1_commit s2_alterdb_with: ERROR:  tuple concurrently updated
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_with s2_begin s2_alterdb_set s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_with: 
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true;
step s2_begin: BEGIN;
step s2_alterdb_set: 
	ALTER DATABASE regress_database SET commit_delay = '10';
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_with s2_begin s2_alterdb_owner s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_with: 
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true;
step s2_begin: BEGIN;
step s2_alterdb_owner: 
	ALTER DATABASE regress_database OWNER TO CURRENT_USER;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_with s2_begin s2_alterdb_tbc s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_with: 
	ALTER DATABASE regress_database WITH ALLOW_CONNECTIONS true;
step s2_begin: BEGIN;
step s2_alterdb_tbc: 
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_set s2_begin s2_alterdb_set s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_set: 
	ALTER DATABASE regress_database SET commit_delay = '10';
step s2_begin: BEGIN;
step s2_alterdb_set: 
	ALTER DATABASE regress_database SET commit_delay = '10'; <waiting ...>
step s1_commit: COMMIT;
step s2_alterdb_set: <... completed>
error in steps s1_commit s2_alterdb_set: ERROR:  duplicate key value violates unique constraint "pg_db_role_setting_databaseid_rol_index"
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_set s2_begin s2_alterdb_owner s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_set: 
	ALTER DATABASE regress_database SET commit_delay = '10';
step s2_begin: BEGIN;
step s2_alterdb_owner: 
	ALTER DATABASE regress_database OWNER TO CURRENT_USER;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_set s2_begin s2_alterdb_tbc s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_set: 
	ALTER DATABASE regress_database SET commit_delay = '10';
step s2_begin: BEGIN;
step s2_alterdb_tbc: 
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_owner s2_begin s2_alterdb_owner s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_owner: 
	ALTER DATABASE regress_database OWNER TO CURRENT_USER;
step s2_begin: BEGIN;
step s2_alterdb_owner: 
	ALTER DATABASE regress_database OWNER TO CURRENT_USER;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_owner s2_begin s2_alterdb_tbc s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_owner: 
	ALTER DATABASE regress_database OWNER TO CURRENT_USER;
step s2_begin: BEGIN;
step s2_alterdb_tbc: 
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_alterdb_tbc s2_begin s2_alterdb_tbc s1_commit s2_commit
step s1_begin: BEGIN;
step s1_alterdb_tbc: 
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT;
step s2_begin: BEGIN;
step s2_alterdb_tbc: 
	ALTER DATABASE regress_database SET TABLESPACE TO DEFAULT;
step s1_commit: COMMIT;
step s2_commit: COMMIT;
